staging_vespalib/src/vespa/vespalib/util/singleexecutor.cpp

```diff
@@ -41,12 +41,38 @@ SingleExecutor::addTask(Task::UP task) {
     return wp;
 }
 
+void
+SingleExecutor::wakeupConsumer() {
+    MonitorGuard guard(_consumerMonitor);
+    guard.signal();
+}
+
+void
+SingleExecutor::sleepConsumer() {
+    _wakeupConsumerAt.store(_wp.load(std::memory_order_relaxed) + (_taskLimit.load(std::memory_order_relaxed) >> 2), std::memory_order_relaxed);
+    MonitorGuard guard(_consumerMonitor);
+    guard.wait(10ms);
+    _wakeupConsumerAt.store(0, std::memory_order_relaxed);
+}
+
+void
+SingleExecutor::wakeupProducer() {
+    MonitorGuard guard(_producerMonitor);
+    guard.signal();
+}
+
+void
+SingleExecutor::sleepProducer(MonitorGuard & guard) {
+    _producerNeedWakeup.store(true, std::memory_order_relaxed);
+    guard.wait(10ms);
+    _producerNeedWakeup.store(false, std::memory_order_relaxed);
+}
+
 Executor::Task::UP
 SingleExecutor::execute(Task::UP task) {
     uint64_t wp = addTask(std::move(task));
     if (wp == _wakeupConsumerAt.load(std::memory_order_relaxed)) {
-        MonitorGuard guard(_consumerMonitor);
-        guard.signal();
+        wakeupConsumer();
```

Comment:
note that this signal may be lost since no state is being passed with the signal (the other thread might call wait after the signal has been discarded since no-one was waiting).

Reply:
Yes, correct. The purpose is to wake up one that is sleeping waiting for low watermark to be reached.